Proposals/20200511-HVPA-design.md

```diff
@@ -0,0 +1,262 @@
+# HVPA-EP-0001: HVPA Redesign
+
+## Summary
+This proposal outlines a plan to enable HVPA to do more efficient scaling horizontally and vertically.
+
+## Motivation
+Currently, HVPA scales by applying user provided weights to HPA and VPA recommendations. We need to provide a better control over scaling of the applications to the users.
+
+### Goals
+
+1. HVPA should be able to scale an application with finer-grain control and predictability over horizontal and vertical scaling which can be specified by the users who are deploying/managing the application.
+
+## Approach
+
+Users will provide buckets of resources each of which would include minimum and maximum CPU, memory and replicas respectively. A configuration is said to belong to a bucket if total resources allocated to all its replcas lie in that bucket. For eg. if there are 3 replicas with 4 cores each, i.e., total 12 cores, then this configuration could belong to a bucket with:
+```
+minimum CPU: 3
+maximum CPU: 6
+minimum memory: 2.5G
+maximum memory: 7G
+minimum replicas: 2
+maximum replicas: 5
+```
+For this bucket mimimum and maximum total CPU are 6 (= 3 * 2) and 30 (= 6 * 5) respectively, and same for the memory would be 5G and 35G respectively.
+So a configuration with 3 replicas with 6GB memory each would also lie in the same bucket.
+
+HVPA controller would calculate the total memory and CPU recommended by the recommenders (eg. HPA or VPA) by multiplying pod resources and the number of replicas, and would find the bucket for the calculated total resources. HVPA controller would then divide the bucket linearly to calculate the actual `replicas` and `resources` to which it will scale the application such that the total allocated resources to all its replicas is same as recommeded by the recommenders.
+If the aggregated recommendations fall in this bucket, then the size of a replica of the scaled application would always be more than the minimum values of resources provided for the bucket, i.e., 3 cores and 2.5G in this case.
+However, the maximum value may not necessarily be bound by its maximum value. For details refer [this section](####base-usage-of-each-replica). For this particular example, if HPA recommends 4 replicas, and VPA recommends 8G memory, then after the scaling, the memory of each replica would be 8GB even though aggregated memory recommendation fall in this bucket.
+
+### User Stories (optional)
+
+#### Story 1
+A user wants to scale both horizontally and vertically with following constraints:
+1. minimum replicas: 1, maximum replicas: 8
+1. per replica minimum memory: 500M, maximum memory: 25G
+1. per replica minimum CPU: 0m, maximum CPU: 24000m
```

Comment:
Can't we hold on to one example, ideally what we want for the API servers? Above is an example with 2-5 replicas, here one with 1-8 replicas, next comes one with 2-8 replicas, but completely different CPU and memory settings.